From 423d042371d0402071c309dc403ea2701600a98b Mon Sep 17 00:00:00 2001
From: nobu <nobu@b2dd03c8-39d4-4d8f-98ff-823fe69b080e>
Date: Sat, 13 Feb 2016 08:12:21 +0000
Subject: [PATCH] no_proxy with whitespaces and leading dots

* lib/uri/generic.rb (find_proxy): exclude white-spaces and allow
  for a leading dot in the domain name in no_proxy.
  [ruby-core:54542] [Feature #8317]

The previous implementation wouldn't allow for white-spaces nor a leading dot
in the domain name. The latter is described in the wget documentation as a valid case.

By being more strict on the characters, which are counted to a domainname,
we allow for white-spaces.
Also, a possible leading dot will be handled gracefully.

[Fix GH-285]

git-svn-id: svn+ssh://ci.ruby-lang.org/ruby/trunk@53816 b2dd03c8-39d4-4d8f-98ff-823fe69b080e
---
 ChangeLog                | 6 ++++++
 lib/uri/generic.rb       | 2 +-
 test/uri/test_generic.rb | 4 ++++
 3 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/ChangeLog b/ChangeLog
index 2945679..44116e0 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,9 @@
+Sat Feb 13 17:11:58 2016  Fabian Wiesel  <fabian.wiesel@sap.com>
+
+	* lib/uri/generic.rb (find_proxy): exclude white-spaces and allow
+	  for a leading dot in the domain name in no_proxy.
+	  [ruby-core:54542] [Feature #8317]
+
 Sat Nov 30 13:28:13 2013  Nobuyoshi Nakada  <nobu@ruby-lang.org>
 
  * siphash.c (sip_hash24): fix for aligned word access little endian
diff --git a/lib/uri/generic.rb b/lib/uri/generic.rb
index aba54c1..f2a2d56 100644
--- a/lib/uri/generic.rb
+++ b/lib/uri/generic.rb
@@ -1662,7 +1662,7 @@ def find_proxy
 
       name = 'no_proxy'
       if no_proxy = ENV[name] || ENV[name.upcase]
-        no_proxy.scan(/([^:,]*)(?::(\d+))?/) {|host, port|
+        no_proxy.scan(/(?!\.)([^:,\s]+)(?::(\d+))?/) {|host, port|
           if /(\A|\.)#{Regexp.quote host}\z/i =~ self.host &&
             (!port || self.port == port.to_i)
             return nil
diff --git a/test/uri/test_generic.rb b/test/uri/test_generic.rb
index fcfe1f9..ad189fc 100644
--- a/test/uri/test_generic.rb
+++ b/test/uri/test_generic.rb
@@ -773,6 +773,14 @@ def test_find_proxy
       assert_equal(URI('http://127.0.0.1:8080'), URI("http://192.0.2.1/").find_proxy)
       assert_nil(URI("http://192.0.2.2/").find_proxy)
     }
+    with_env('http_proxy'=>'http://127.0.0.1:8080', 'no_proxy'=>'example.org') {
+      assert_nil(URI("http://example.org/").find_proxy)
+      assert_nil(URI("http://www.example.org/").find_proxy)
+    }
+    with_env('http_proxy'=>'http://127.0.0.1:8080', 'no_proxy'=>'.example.org') {
+      assert_nil(URI("http://example.org/").find_proxy)
+      assert_nil(URI("http://www.example.org/").find_proxy)
+    }
     with_env('http_proxy'=>'') {
       assert_nil(URI("http://192.0.2.1/").find_proxy)
       assert_nil(URI("ftp://192.0.2.1/").find_proxy)
